File: VB/DiagramContainersVB/Controllers/HomeController.vb
Imports DevExpress.Web.Mvc

Public Class HomeController
    Inherits System.Web.Mvc.Controller

    Function Index() As ActionResult
        Return View(DiagramDataProvider.GetItems())
    End Function

    Public Function DiagramUpdate(ByVal nodeUpdateValues As MVCxDiagramNodeUpdateValues(Of Item, Integer)) As ActionResult
        For Each item In nodeUpdateValues.Update
            DiagramDataProvider.UpdateItem(item)
        Next

        For Each itemKey In nodeUpdateValues.DeleteKeys
            DiagramDataProvider.DeleteItem(itemKey)
        Next

        For Each item In nodeUpdateValues.Insert
            Dim key = DiagramDataProvider.InsertItem(item)
            nodeUpdateValues.MapInsertedItemKey(item, key)
        Next

        Return DiagramExtension.GetBatchUpdateResult(nodeUpdateValues)
    End Function
End Class

File: CS/DiagramContainers/Controllers/HomeController.cs
using DevExpress.Web.Mvc;
using DiagramContainers.Models;
using System.Web.Mvc;

namespace DiagramContainers.Controllers
{
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            return View(DiagramDataProvider.GetItems());
        }

        public ActionResult DiagramUpdate(MVCxDiagramNodeUpdateValues<Item, int> nodeUpdateValues)
        {
            foreach (var item in nodeUpdateValues.Update)
            {
                DiagramDataProvider.UpdateItem(item);
            }
            foreach (var itemKey in nodeUpdateValues.DeleteKeys)
            {
                DiagramDataProvider.DeleteItem(itemKey);
            }
            foreach (var item in nodeUpdateValues.Insert)
            {
                var key = DiagramDataProvider.InsertItem(item);
                nodeUpdateValues.MapInsertedItemKey(item, key);
            }

            return DiagramExtension.GetBatchUpdateResult(nodeUpdateValues);
        }
    }
}